### PR TITLE
feat: make sharding configurable in wrap_dataset

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
Users have asked for the ability to perform dataset sharding themselves. Previously this was only supported via `data: dataflow_to_tf_dataset` in EstimatorTrial which was a hack.

### CODE REVIEW

**Comment by aaron276h** [comment]
@rb-determined-ai If we were to make it optional, unless users read our documentation very carefully most would be unaware that this call needs to take place (which would lead to incorrect behaviour in many uses cases) and we would have no method of checking / enforcing correct behaviour.

**Comment by rb-determined-ai** [comment]
@aaron276h while I don't disagree with what you are saying, it still feels like a very silly API.  Is there a different approach we can take here?

For example, the name `wrap_dataset` is opaque in it's meaning to the user.  That name makes sense if one of the following cases is true:
1. the user is coming from a non-dtrain paradigm and does not really understand or expect sharding
2. we are reserving the right to do arbitrary transformations on the dataset

Otherwise it might make sense to rename `wrap_dataset` as `shard_dataset`, since that's what it really is.  I would argue that we've gone far enough down this road that we already know that (2) is not the case.

So what if we offered two calls, one was `context.shard_dataset(ds)` and one was `context.disable_shard_dataset_check()` or something?

I'm not saying that I particularly like the solution I just proposed, but what I'm saying is that I think offering `wrap_dataset(shard=False)` is really just an API bandaid, and I think we should aim for a generally good API rather than patch up the old one.  I don't think this is a time-critical issue because the one user we know of with this issue can call `context.wrap_dataset(tf.data.Dataset.range(1))` to disable the check in the mean time.

(Separately, do we still need the random seed differences that come with `dataflow_to_dataset=true`?)

**Comment by aaron276h** [comment]
@rb-determined-ai for the random seed I want to say no, but also not 100% convinced of this.

I agree that `shard_dataset()` is a better name, but don't think that this rename warrants a breaking change which would impact all tf_keras and estimator models.

**Comment by rb-determined-ai** [comment]
@aaron276h I don't see why this needs to be a breaking change.  It is trivially easy to let `wrap_dataset` be a deprecated alias for `shard_dataset`.

It seems we should talk to some AMLEs and SEs and try to come up with a good API here.  Do you think `shard_dataset()` is the best we can do?

File: harness/determined/_experiment_config.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def mixed_precision_enabled(self) -> bool:
     def averaging_training_metrics_enabled(self) -> bool:
         return bool(self["optimizations"]["average_training_metrics"])
 
+    # TODO (DET-3798): Remove input_from_dataflow in 0.12.14.
```
Comment (rb-determined-ai) [inline harness/determined/_experiment_config.py:30]: No need to put a hard version on this.  We can just have the comment say "remove this after known customers have migrated" or just "remove this at some point" and put details in the jira ticket.

File: harness/determined/estimator/_estimator_context.py [hunk 2]
```diff
@@ -82,12 +83,23 @@ def wrap_dataset(self, dataset: Any) -> Any:
         independently. E.g., If users instantiate their training dataset within
         ``build_train_spec()``, they should call ``dataset = wrap_dataset(dataset)``
         prior to passing it into ``tf.estimator.TrainSpec``.
+
+        Args:
+            dataset: tf.data.Dataset
+            shard_dataset: When performing multi-slot (distributed) training, this
```
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_context.py:89]: question: does this syntax render properly as restructured text?  I think in other places we use a hanging indent to distinguish between arguments and their descriptions.

**Comment by rb-determined-ai** [comment]
@aaron276h actually, can we just add this argument to TFKerasContext.wrap_dataset as well while we are at it?

**Comment by aaron276h** [comment]
@rb-determined-ai this PR includes the changes for TFKeras as well.

File: harness/determined/_experiment_config.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/_experiment_config.py]: done

**Comment by rb-determined-ai** [comment]
Ah, yeah I just missed that detail.  Nice work : )

File: harness/determined/estimator/_estimator_context.py (hunk 2 above)
Comment (aaron276h) [reply harness/determined/estimator/_estimator_context.py]: really good call, turns out the way I had it was not formatting correctly. Should be fixed now.

**Pull request merged**